connectors/sources/slack.py

```diff
@@ -0,0 +1,393 @@
+            yield message, None
+
+    async def channels_and_messages(self):
+        delta = timedelta(days=self.n_days_to_fetch)
+        past_unix_timestamp = time.mktime((datetime.utcnow() - delta).timetuple())
+        current_unix_timestamp = time.mktime(datetime.utcnow().timetuple())
+        async for channel in self.slack_client.list_channels(
+            not self.auto_join_channels
+        ):
+            self._logger.info(f"Listed channel: '{channel['name']}'")
+            yield self.remap_channel(channel)
+            channel_id = channel["id"]
+            join_response = {}
+            if self.auto_join_channels:
+                join_response = await self.slack_client.join_channel(
+                    channel_id
+                )  # can't get channel content if the bot is not in the channel
+            if not self.auto_join_channels or join_response.get("ok"):
+                async for message in self.slack_client.list_messages(
+                    channel, past_unix_timestamp, current_unix_timestamp
+                ):
+                    yield self.remap_message(message, channel)
+            else:
+                self._logger.warning(
+                    f"Not syncing channel: '{channel['name']}' because: {join_response.get('error')}"
+                )
+
+    def get_username(self, user):
+        """
+        Given a user record from slack, try to find a good username for it.
+        This is hard, because no one property is reliably present and optimal.
+        This function goes through a variety of options, from most normalized and readable to least.
+        :param user:
+        :return: an identifier for the user - hopefully a human-readable name.
+        """
+        user_profile = user.get("profile", {})
+        if user_profile.get("display_name_normalized"):
+            return user_profile["display_name_normalized"]
+        elif user_profile.get("real_name_normalized"):
+            return user_profile["real_name_normalized"]
+        elif user.get("real_name"):
+            return user["real_name"]
+        elif user.get("name"):
+            return user["name"]
+        else:
+            return user[
+                "id"
+            ]  # Some Users do not have any names (like Bots). For these, we fall back on ID
+
+    def remap_user(self, user):
+        user_profile = user.get("profile", {})
+        return {
+            "_id": user["id"],
+            "type": "user",
+            "real_name_normalized": user_profile.get("real_name_normalized"),
+            "profile_real_name": user_profile.get("real_name"),
+            "display_name": user_profile.get("display_name"),
+            "display_name_normalized": user_profile.get("display_name_normalized"),
+            "first_name": user_profile.get("first_name"),
+            "last_name": user_profile.get("last_name"),
+        } | dict_slice(
+            user,
+            (
+                "id",
```

Comment:
Is it duplicated?

Reply:
yes, both in `id` and in `_id`. But this is intentional, so that an ES response of _source still contains the relevant slack metadata.

Reply:
In Extractor, the `_id` is poped and stored as `id`:

https://github.com/elastic/connectors-python/blob/a03cc1c9ba9c0d086ccd079b49c24eec5ec86c79/connectors/es/sink.py#L392C17-L392C52

So the `id` here will be eventually overwritten by `_id`, which is essentially, a no-op.

This is happening today. For any connector, you can check the documents in the index `search-*`, and you will find duplicated `id` field and document id `_id`. I don't feel this is right. Just FYI